File: algorithms/2_medium/M/03499_maximize_active_section_with_trade_i/solution.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static int maxActiveSectionsAfterTrade(const string_view s) {
        const isize len = std::ssize(s);
        isize total_ones = 0;
        isize prev_zeros = 0;
        isize curr_zeros = 0;
        isize max = 0;

        for (isize i = 0; i < len; ++i) {
            switch (s[i]) {
                case '0':
                    curr_zeros += 1;
                    if (prev_zeros > 0) {
                        max = std::max(max, prev_zeros + curr_zeros);
                    }
                    break;
                case '1':
                    if (i > 0 && s[i - 1] == '0') {
                        prev_zeros = curr_zeros;
                        curr_zeros = 0;
                    }
                    total_ones += 1;
                    break;
                default:
                    break;
            }
        }

        assert(max + total_ones <= numeric_limits<int>::max());
        return max + total_ones;
    }
};

File: algorithms/2_medium/M/03499_maximize_active_section_with_trade_i/solution.rs
impl Solution {
    pub fn max_active_sections_after_trade(s: String) -> i32 {
        let bytes = s.as_bytes();
        let (mut prev, mut curr, mut ones, mut max) = (0_isize, 0_isize, 0_isize, 0_isize);

        for (i, &b) in bytes.iter().enumerate() {
            match b {
                b'1' => {
                    ones += 1;
                    if i > 0 && bytes[i - 1] == b'0' {
                        (prev, curr) = (curr, 0);
                    }
                },
                b'0' => {
                    curr += 1;
                    if prev > 0 && curr > 0 {
                        max = max.max(curr + prev);
                    }
                },
                _ => unreachable!("invalid input"),
            }
        }

        (max + ones).try_into().unwrap()
    }
}
